ocaml/lambda/translcore.ml

```diff
@@ -1004,18 +1009,21 @@ and transl_apply ~scopes
     | Omitted { mode_closure; mode_arg; mode_ret } :: l ->
         assert (pos = Rc_normal);
         let defs = ref [] in
-        let protect name lam =
+        let protect name (lam, layout) =
           match lam with
-            Lvar _ | Lconst _ -> lam
+            Lvar _ | Lconst _ -> (lam, layout)
           | _ ->
               let id = Ident.create_local name in
-              defs := (id, lam) :: !defs;
-              Lvar id
+              defs := (id, layout, lam) :: !defs;
+              (Lvar id, layout)
         in
         let lam =
-          if args = [] then lam else lapply lam (List.rev args) loc pos ap_mode
+          if args = [] then
+            lam
+          else
+            lapply lam (List.rev args) loc pos ap_mode layout_function
```

Comment:
I am slightly surprised to see `layout_function` here, which I imagined to be a place-holder for the layout of functions, not return values.

Reply:
In this case, `lam` is either the original function (`args = []`) or a partial application of the original function to its first arguments, up to (and excluding) the current one. In both cases this should correspond to a functional value.